src/orchestrator.ts

```diff
@@ -562,20 +569,22 @@ export class Orchestrator {
     }
 
     private any(state: HistoryEvent[], tasks: TaskBase[]): TaskSet {
-        const completedTasks = tasks
-            .filter(TaskFilter.isCompletedTask)
-            .sort((a, b) => {
-                if (a.completionIndex > b.completionIndex) { return 1; }
-                if (a.completionIndex < b.completionIndex) { return -1; }
-                return 0;
-            });
+        if (!tasks || tasks.length === 0) {
+            throw new Error("At least one yieldable task must be provided to wait for.");
+        }
 
-        const firstCompleted = completedTasks[0];
-        if (firstCompleted) {
-            return TaskFactory.SuccessfulTaskSet(tasks, firstCompleted.completionIndex, firstCompleted);
-        } else {
+        const completedTasks = tasks.filter(TaskFilter.isCompletedTask);
+        if (completedTasks.length === 0) {
             return TaskFactory.UncompletedTaskSet(tasks);
         }
+
+        // Find the completed task with the lowest completion index
+        let firstCompleted = completedTasks[0];
+        completedTasks.forEach((task) =>
+            firstCompleted = task.completionIndex < firstCompleted.completionIndex ? task : firstCompleted,
+        );
```

Comment:
It looks like we're iterating through the list of tasks twice. Should we try to reduce it to one iteration?

```typescript
let firstCompleted: TaskBase = null;
completedTasks.forEach((task) => {
    if (TaskFilter.isCompletedTask(task)) {
        if (!firstCompleted || task.completionIndex < firstCompleted.completionIndex) {
            firstCompleted = task;
        }
    }
});

if (firstCompleted) {
    return TaskFactory.SuccessfulTaskSet(tasks, firstCompleted.completionIndex, firstCompleted);
} else {
    return TaskFactory.UncompletedTaskSet(tasks);
}
```

Reply:
There is definitely room to reduce iteration through tasks. I actually think task.all() is the worse offender here. I am going to try to optimize both of these, as the fan-out scenario can cause the internal task arrays to get rather large.